File: src/applications/arm_demos/scripts/mtc_simple_demo.py
#!/usr/bin/env python3
"""
MoveIt Task Constructor Simple Demo

A simplified demonstration of MTC showing basic task composition:
- Move to home position
- Approach a target pose
- Retreat back
- Return to home

This example is ideal for learning MTC concepts without gripper complexity.

Author: LDR Robotics Team
License: MIT
"""

import sys

# Check if MTC is available
try:
    from moveit.task_constructor import core, stages
except ImportError as e:
    print("\n" + "="*60)
    print("ERROR: MoveIt Task Constructor not installed!")
    print("="*60)
    print(f"\nImport error: {e}")
    print("\nTo install MTC, run:")
    print("  sudo apt install ros-humble-moveit-task-constructor-*")
    print("\nFor detailed installation instructions, run:")
    print("  ros2 run arm_demos check_mtc.py")
    print("\n" + "="*60)
    sys.exit(1)

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped, Vector3Stamped


class MTCSimpleDemo(Node):
    """Simple MTC demonstration node."""

    def __init__(self):
        super().__init__('mtc_simple_demo')

        self.task = None
        self.arm_group_name = "arm"
        self.hand_frame = "wrist_roll_link"

        self.get_logger().info("MTC Simple Demo initialized")

    def create_task(self):
        """Create a simple MTC task."""
        self.task = core.Task("simple_approach_retreat")

        # Load robot model
        self.task.loadRobotModel(self.get_logger())

        # Set task properties
        self.task.setProperty("group", self.arm_group_name)
        self.task.setProperty("ik_frame", self.hand_frame)

        # ==========================
        # Create planners
        # ==========================

        # Sampling planner (OMPL)
        sampling_planner = core.PipelinePlanner()
        sampling_planner.setProperty("goal_joint_tolerance", 1e-5)

        # Cartesian planner
        cartesian_planner = core.CartesianPath()
        cartesian_planner.setMaxVelocityScalingFactor(0.5)  # Slower for safety
        cartesian_planner.setMaxAccelerationScalingFactor(0.5)
        cartesian_planner.setStepSize(0.01)

        # ==========================
        # Define stages
        # ==========================

        # Stage 1: Start from current state
        current_state = stages.CurrentState("current state")
        self.task.add(current_state)

        # Stage 2: Move to home position
        move_to_home = stages.MoveTo("move to home", sampling_planner)
        move_to_home.setGroup(self.arm_group_name)
        move_to_home.setGoal("home")  # Named pose from SRDF
        self.task.add(move_to_home)

        # Stage 3: Move forward (approach)
        approach = stages.MoveRelative("approach", cartesian_planner)
        approach.setGroup(self.arm_group_name)
        approach.setMinMaxDistance(0.10, 0.20)  # Move 10-20cm

        # Set direction (forward in X)
        direction_forward = Vector3Stamped()
        direction_forward.header.frame_id = self.hand_frame
        direction_forward.vector.x = 1.0  # Forward
        approach.setDirection(direction_forward)

        self.task.add(approach)

        # Stage 4: Retreat (move back)
        retreat = stages.MoveRelative("retreat", cartesian_planner)
        retreat.setGroup(self.arm_group_name)
        retreat.setMinMaxDistance(0.10, 0.20)

        # Set direction (backward in X)
        direction_backward = Vector3Stamped()
        direction_backward.header.frame_id = self.hand_frame
        direction_backward.vector.x = -1.0  # Backward
        retreat.setDirection(direction_backward)

        self.task.add(retreat)

        # Stage 5: Return to home
        return_home = stages.MoveTo("return to home", sampling_planner)
        return_home.setGroup(self.arm_group_name)
        return_home.setGoal("home")
        self.task.add(return_home)

        self.get_logger().info("Simple MTC task created")

    def plan_task(self):
        """Plan the task."""
        if self.task is None:
            self.get_logger().error("Task not created")
            return False

        self.get_logger().info("Planning task...")

        try:
            self.task.plan()

            num_solutions = self.task.numSolutions()
            if num_solutions == 0:
                self.get_logger().error("No solutions found!")
                return False

            self.get_logger().info(f"Found {num_solutions} solution(s)")

            # Print the best solution
            best_solution = self.task.solutions()[0]
            self.get_logger().info(f"Best solution cost: {best_solution.cost()}")

            return True

        except Exception as e:
            self.get_logger().error(f"Planning failed: {str(e)}")
            return False

    def execute_task(self):
        """Execute the planned task."""
        if self.task is None or self.task.numSolutions() == 0:
            self.get_logger().error("No solutions to execute")
            return False

        self.get_logger().info("Executing task...")

        try:
            # Execute the best solution
            result = self.task.execute(self.task.solutions()[0])
            self.get_logger().info("Execution completed successfully!")
            return True

        except Exception as e:
            self.get_logger().error(f"Execution failed: {str(e)}")
            return False

    def print_task_structure(self):
        """Print the task structure for debugging."""
        if self.task is None:
            self.get_logger().error("Task not created")
            return

        self.get_logger().info("\n" + "="*50)
        self.get_logger().info("Task Structure:")
        self.get_logger().info("="*50)
        self.get_logger().info(self.task.toString())
        self.get_logger().info("="*50 + "\n")


def main(args=None):
    """Main entry point."""
    rclpy.init(args=args)

    demo = MTCSimpleDemo()

    try:
        # Create the task
        demo.create_task()

        # Print task structure
        demo.print_task_structure()

        # Plan the task
        if demo.plan_task():
            # Prompt for execution
            print("\n" + "="*50)
            print("Planning succeeded!")
            print("="*50)
            response = input("\nExecute the task? (y/n): ")

            if response.lower() == 'y':
                demo.execute_task()
            else:
                demo.get_logger().info("Execution skipped by user")
        else:
            demo.get_logger().error("Planning failed. Cannot execute.")

    except KeyboardInterrupt:
        demo.get_logger().info("Interrupted by user")
    except Exception as e:
        demo.get_logger().error(f"Error: {str(e)}")
    finally:
        demo.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/applications/arm_demos/src/mtc_simple_demo.cpp
/**
 * @file mtc_simple_demo.cpp
 * @brief MoveIt Task Constructor Simple Demo (C++ port)
 *
 * A simplified demonstration of MTC showing basic task composition:
 * - Move to home position
 * - Approach a target pose
 * - Retreat back
 * - Return to home
 *
 * This example is ideal for learning MTC concepts without gripper complexity.
 *
 * Author: LDR Robotics Team
 * License: MIT
 */

#include <rclcpp/rclcpp.hpp>
#include <moveit/task_constructor/task.h>
#include <moveit/task_constructor/solvers.h>
#include <moveit/task_constructor/stages.h>

#if __has_include(<tf2_geometry_msgs/tf2_geometry_msgs.hpp>)
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#else
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#endif

#include <geometry_msgs/msg/vector3_stamped.hpp>

namespace mtc = moveit::task_constructor;

static const rclcpp::Logger LOGGER = rclcpp::get_logger("mtc_simple_demo");

class MTCSimpleDemo
{
public:
  MTCSimpleDemo(const rclcpp::NodeOptions& options);

  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr getNodeBaseInterface();

  void run();

private:
  mtc::Task createTask();
  mtc::Task task_;
  rclcpp::Node::SharedPtr node_;

  std::string arm_group_name_;
  std::string hand_frame_;
};

MTCSimpleDemo::MTCSimpleDemo(const rclcpp::NodeOptions& options)
  : node_(std::make_shared<rclcpp::Node>("mtc_simple_demo", options))
{
  arm_group_name_ = "arm";
  hand_frame_ = "wrist_roll_link";
  RCLCPP_INFO(LOGGER, "MTC Simple Demo initialized");
}

rclcpp::node_interfaces::NodeBaseInterface::SharedPtr
MTCSimpleDemo::getNodeBaseInterface()
{
  return node_->get_node_base_interface();
}

mtc::Task MTCSimpleDemo::createTask()
{
  mtc::Task task;
  task.stages()->setName("simple_approach_retreat");
  task.loadRobotModel(node_);

  // Set task properties
  task.setProperty("group", arm_group_name_);
  task.setProperty("ik_frame", hand_frame_);

  // Create planners
  auto sampling_planner = std::make_shared<mtc::solvers::PipelinePlanner>(node_);

  auto cartesian_planner = std::make_shared<mtc::solvers::CartesianPath>();
  cartesian_planner->setMaxVelocityScalingFactor(0.5);
  cartesian_planner->setMaxAccelerationScalingFactor(0.5);
  cartesian_planner->setStepSize(0.01);

  // Stage 1: Start from current state
  {
    auto stage = std::make_unique<mtc::stages::CurrentState>("current state");
    task.add(std::move(stage));
  }

  // Stage 2: Move to home position
  {
    auto stage = std::make_unique<mtc::stages::MoveTo>("move to home", sampling_planner);
    stage->setGroup(arm_group_name_);
    stage->setGoal("home");
    task.add(std::move(stage));
  }

  // Stage 3: Move forward (approach)
  {
    auto stage = std::make_unique<mtc::stages::MoveRelative>("approach", cartesian_planner);
    stage->setGroup(arm_group_name_);
    stage->setMinMaxDistance(0.10, 0.20);

    geometry_msgs::msg::Vector3Stamped direction;
    direction.header.frame_id = hand_frame_;
    direction.vector.x = 1.0;  // Forward
    stage->setDirection(direction);
    task.add(std::move(stage));
  }

  // Stage 4: Retreat (move back)
  {
    auto stage = std::make_unique<mtc::stages::MoveRelative>("retreat", cartesian_planner);
    stage->setGroup(arm_group_name_);
    stage->setMinMaxDistance(0.10, 0.20);

    geometry_msgs::msg::Vector3Stamped direction;
    direction.header.frame_id = hand_frame_;
    direction.vector.x = -1.0;  // Backward
    stage->setDirection(direction);
    task.add(std::move(stage));
  }

  // Stage 5: Return to home
  {
    auto stage = std::make_unique<mtc::stages::MoveTo>("return to home", sampling_planner);
    stage->setGroup(arm_group_name_);
    stage->setGoal("home");
    task.add(std::move(stage));
  }

  RCLCPP_INFO(LOGGER, "Simple MTC task created");
  return task;
}

void MTCSimpleDemo::run()
{
  RCLCPP_INFO(LOGGER, "Creating task...");
  task_ = createTask();

  // Initialize task
  try {
    task_.init();
    RCLCPP_INFO(LOGGER, "Task initialized successfully");
  } catch (mtc::InitStageException& e) {
    RCLCPP_ERROR_STREAM(LOGGER, "Task initialization failed: " << e);
    return;
  }

  // Plan task
  RCLCPP_INFO(LOGGER, "Planning task...");
  if (!task_.plan(5)) {
    RCLCPP_ERROR(LOGGER, "Task planning failed - no solutions found");
    return;
  }

  RCLCPP_INFO(LOGGER, "Planning succeeded! %zu solution(s) found", task_.solutions().size());

  // Publish solution for visualization
  task_.introspection().publishSolution(*task_.solutions().front());

  // Execute the best solution
  RCLCPP_INFO(LOGGER, "Executing task...");
  auto result = task_.execute(*task_.solutions().front());
  if (result.val != moveit_msgs::msg::MoveItErrorCodes::SUCCESS) {
    RCLCPP_ERROR(LOGGER, "Task execution failed with error code: %d", result.val);
    return;
  }

  RCLCPP_INFO(LOGGER, "Execution completed successfully!");
}

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);

  rclcpp::NodeOptions options;
  options.automatically_declare_parameters_from_overrides(true);

  auto demo = std::make_shared<MTCSimpleDemo>(options);

  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(demo->getNodeBaseInterface());

  auto spin_thread = std::make_unique<std::thread>([&executor]() {
    executor.spin();
  });

  // Wait for joint states to be available
  RCLCPP_INFO(LOGGER, "Waiting for system to be ready...");
  rclcpp::sleep_for(std::chrono::seconds(3));

  demo->run();

  RCLCPP_INFO(LOGGER, "Demo completed. Keeping node alive for visualization. Press Ctrl+C to exit.");
  spin_thread->join();
  rclcpp::shutdown();
  return 0;
}
